### ISSUE TITLE
LLL reduction - ld error

### ISSUE THREAD

**Issue by johughes99** [fplll/fplll#415]
I trying to do a LLL reduction - based on a simple example (see attached).

I'm following the LLL test example and making the following call:

status = lll_reduction(lattice, LLL_DEF_DELTA, LLL_DEF_ETA, LM_PROVED, FT_MPFR, 0, LLL_DEFAULT);

This compiles ok - but I'm getting the following  link error:

/tmp/cc2W7RS7.o: In function `main':
fplll_example.cpp:(.text+0x444): undefined reference to `fplll::lll_reduction(fplll::ZZ_mat<__mpz_struct [1]>&, double, double, fplll::LLLMethod, fplll::FloatType, int, int)'
collect2: error: ld returned 1 exit status

This implies I'm not using the correct call to the LLL reduction method.

[fplll_example.txt](https://github.com/fplll/fplll/files/4443188/fplll_example.txt)

**Comment by malb** [comment]
Works for me:

```
gcc -I$VIRTUAL_ENV/include/ -lstdc++ 'pkg-config fplll --libs' fplll_example.cpp -o fplll_example
```

**Comment by johughes99** [comment]
Still no joy. I have made sure that mpfr, gmp and fplll are all compiled and linked with gcc 8.4

Tried all combination - buts still get undefined reference errors

What do you have as `pkg-config fplll --libs` set up to be?

If I try:

      g++-8 -lstdc++ -I/usr/local/include/fplll -lfplll -lgmp fplll_example.cpp -o fplll 

I get the following link errors - which implies it cant find either th gmp or fplll symbols

fplll_example.cpp:(.text+0x3a6): undefined reference to `__gmpz_init'
fplll_example.cpp:(.text+0x3b7): undefined reference to `__gmpz_set_ui'
fplll_example.cpp:(.text+0x44b): undefined reference to `fplll::lll_reduction(fplll::ZZ_mat<__mpz_struct [1]>&, double, double, fplll::LLLMethod, fplll::FloatType, int, int)'
/tmp/ccVDNgLu.o: In function `fplll::Z_NR<__mpz_struct [1]>::Z_NR()':
fplll_example.cpp:(.text._ZN5fplll4Z_NRIA1_12__mpz_structEC2Ev[_ZN5fplll4Z_NRIA1_12__mpz_structEC5Ev]+0x14): undefined reference to `__gmpz_init'
/tmp/ccVDNgLu.o: In function `fplll::Z_NR<__mpz_struct [1]>::Z_NR(fplll::Z_NR<__mpz_struct [1]> const&)':
fplll_example.cpp:(.text._ZN5fplll4Z_NRIA1_12__mpz_structEC2ERKS3_[_ZN5fplll4Z_NRIA1_12__mpz_structEC5ERKS3_]+0x1f): undefined reference to `__gmpz_init_set'
/tmp/ccVDNgLu.o: In function `fplll::Z_NR<__mpz_struct [1]>::~Z_NR()':
fplll_example.cpp:(.text._ZN5fplll4Z_NRIA1_12__mpz_structED2Ev[_ZN5fplll4Z_NRIA1_12__mpz_structED5Ev]+0x14): undefined reference to `__gmpz_clear'
/tmp/ccVDNgLu.o: In function `fplll::Z_NR<__mpz_struct [1]>::operator=(__mpz_struct const (&) [1])':
fplll_example.cpp:(.text._ZN5fplll4Z_NRIA1_12__mpz_structEaSERA1_KS1_[_ZN5fplll4Z_NRIA1_12__mpz_structEaSERA1_KS1_]+0x1f): undefined reference to `__gmpz_set'
/tmp/ccVDNgLu.o: In function `fplll::Z_NR<__mpz_struct [1]>::operator=(long)':
fplll_example.cpp:(.text._ZN5fplll4Z_NRIA1_12__mpz_structEaSEl[_ZN5fplll4Z_NRIA1_12__mpz_structEaSEl]+0x1f): undefined reference to `__gmpz_set_si'

**Comment by johughes99** [comment]
Tried lots of combinations and different orders of linking - but still get the above link errors.

Tried:

    g++-8 -lstdc++ -lgmp -lmpfr -lfplll -I/usr/local/include/fplll -I/usr/local/include -L/usr/local/lib fplll_example.cpp -o fplll_test

which is basically what worked before.

It comes down to the fact that as soon as I add in the Line

`status = lll_reduction(lattice, LLL_DEF_DELTA, LLL_DEF_ETA, LM_PROVED, FT_MPFR, 0, LLL_DEFAULT);`

I start getting these link errors - so all the other calls to fplll seem to cause no problems.

**Comment by malb** [comment]
Maybe you have an old and a new FPLLL installation all mixed up or so?

**Comment by malb** [comment]
I get

```sh
$ pkg-config fplll --libs
-L/BLABLABLA/lib -lpthread -lqd -lmpfr -lgmp -lfplll
```

**Comment by johughes99** [comment]
Good idea.  Let me delete all the .a and .h installed files and rebuild and reinstall from scratch  - for mpfr, gmp and fplll

**Comment by johughes99** [comment]
And of course do a make clean before doing the actual make.

**Comment by johughes99** [comment]
I rebuilt etc and when I came to compile/link the example build I still got the same errors.  At that point I changed tack - as errors like this quite often are due to linking order errors. And this is despite I was following the order provided above.

After a lot of testing I found only one combination that worked. and this involved trying both -l and explicit library path testing....

The combination and order was

`g++-8 -lstdc++ -pthread -I/usr/local/include/fplll -I/usr/local/include/ fplll_example.cpp -o fplll_test  /usr/local/lib/libfplll.a   /usr/local/lib/libmpfr.a  /usr/local/lib/libgmp.a

I tried various combination and order of -lgmp, -lmpfr and -lfplll as well - but they all failed with link errors - with either a few - or a lot of errors.

So quite mystified why the above worked - but using the -l approach didn't

At least I have a solution - and I can continue to make progress
`
